samples/tutorial/Tutorial4.md

```diff
@@ -0,0 +1,343 @@
+
+```kotlin
+object TodoListWorkflow : StatefulWorkflow<ListProps, Unit, Output, TodoListScreen>() {
+
+  data class ListProps(
+    val name: String,
+    val todos: List<TodoModel>
+  )
+
+  sealed class Output {
+    object Back : Output()
+    data class SelectTodo(val index: Int) : Output()
+    object NewTodo : Output()
+  }
+
+  override fun initialState(
+    props: ListProps,
+    snapshot: Snapshot?
+  ) = Unit
+
+  // …
+}
+```
+
+TODO add section about converting to StatelessWorkflow.
+
+Change the `WorkflowAction` behaviors to return an output instead of modifying any state:
+
+```kotlin
+object TodoListWorkflow : StatefulWorkflow<ListProps, Unit, Output, TodoListScreen>() {
+
+  // …
+
+  private fun onBack() = action {
+    // When an onBack action is received, emit a Back output.
+    setOutput(Back)
+  }
+
+  private fun selectTodo(index: Int) = action {
+    // Tell our parent that a todo item was selected.
+    setOutput(SelectTodo(index))
+  }
+
+  private fun new() = action {
+    // Tell our parent a new todo item should be created.
+    setOutput(NewTodo)
+  }
+}
+```
+
+Update the `render` method to only return the `TodoListScreen`:
+
+```kotlin
+object TodoListWorkflow : StatefulWorkflow<ListProps, Unit, Output, TodoListScreen>() {
+
+  // …
+
+  override fun render(
+    props: ListProps,
+    state: Unit,
+    context: RenderContext
+  ): TodoListScreen {
+    val titles = props.todos.map { it.title }
+    return TodoListScreen(
+        name = props.name,
+        todoTitles = titles,
+        onTodoSelected = { context.actionSink.send(selectTodo(it)) },
+        onBack = { context.actionSink.send(onBack()) }
+    )
+  }
+
+  // …
+
+}
+```
+
+Render the `TodoListWorkflow` and handle its output in the `TodoWorkflow`:
+
+```kotlin
+object TodoWorkflow : StatefulWorkflow<TodoProps, State, Back, List<Any>>() {
+
+  // …
+
+  override fun render(
+    props: TodoProps,
+    state: State,
+    context: RenderContext
+  ): List<Any> {
+    val todoListScreen = context.renderChild(
+        TodoListWorkflow,
+        props = ListProps(
+            name = props.name,
+            todos = state.todos
+        )
+    ) { output ->
+      when (output) {
+        Output.Back -> onBack()
+        is SelectTodo -> editTodo(output.index)
+        NewTodo -> newTodo()
+      }
+    }
+
+    return listOf(todoListScreen)
+  }
+
+  private fun onBack() = action {
+    // When an onBack action is received, emit a Back output.
+    setOutput(Back)
+  }
+
+  private fun editTodo(index: Int) = action {
+    // When a todo item is selected, edit it.
+    state = state.copy(step = Step.Edit(index))
+  }
+
+  private fun newTodo() = action {
+    // Append a new todo model to the end of the list.
+    state = state.copy(
+        todos = state.todos + TodoModel(
+            title = "New Todo",
+            note = ""
```

Comment:
Nit: May want to add a note to say that we have no separate screen for adding a todo, we just fill out a default one and drop it into the list (i.e. I was expecting us to change to Step.Edit) here, but we didn't.